File: Data Structures & Algorithms/add-two-numbers/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
              
# l1=[9,9,9,9,9,9,9]
# l2=[9,9,9,9]
#     8 9 9 9 

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        dummy = curr = ListNode()
        carry = None
        l1_val, l2_val = None, None

        while l1 != None or l2 != None:



            if l1:
                l1_val = l1.val
                l1 = l1.next
            else:
                l1_val = 0

            if l2:
                l2_val = l2.val
                l2 = l2.next
            else:
                l2_val = 0
            
            val = l1_val + l2_val

            if carry:
                val += 1
                carry = None

            if val >= 10:
                carry = 1
                val %= 10
            
            curr.next = ListNode(val=val)
            curr = curr.next

        if carry:
            curr.next = ListNode(val = 1)

        return dummy.next

            

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode dummy(0, head);

        ListNode* l = &dummy;
        for (int i = 1; i < left; i++) {
            l = l->next;
        }

        ListNode* r = &dummy;
        for (int i = 0; i < right; i++) {
            r = r->next;
        }

        ListNode* temp = r->next;

        ListNode* curr = l->next;
        ListNode* end = curr;

        ListNode* tail = nullptr;

        while (curr != temp) {
            ListNode* next = curr->next;

            curr->next = tail;
            tail = curr;

            curr = next;
        }

        l->next = tail;
        end->next = temp;

        return dummy.next;
    }
};